src/distPlus/distribution/Continuous.re

```diff
@@ -36,59 +51,89 @@ let combinePointwise =
   // can just sum them up. Otherwise, all bets are off.
   let combinedIntegralSum =
     Common.combineIntegralSums(
-      knownIntegralSumsFn,
-      t1.knownIntegralSum,
-      t2.knownIntegralSum,
+      integralSumCachesFn,
+      t1.integralSumCache,
+      t2.integralSumCache,
     );
 
+  // TODO: does it ever make sense to pointwise combine the integrals here?
+  // It could be done for pointwise additions, but is that ever needed?
+
+  // If combining stepwise and linear, we must convert the stepwise to linear first,
+  // i.e. add a point at the bottom of each step
+  let (t1, t2) = switch (t1.interpolation, t2.interpolation) {
+  | (`Linear, `Linear) => (t1, t2);
+  | (`Stepwise, `Stepwise) => (t1, t2);
+  | (`Linear, `Stepwise) => (t1, stepwiseToLinear(t2));
+  | (`Stepwise, `Linear) => (stepwiseToLinear(t1), t2);
+  };
+
+  let interpolator = XYShape.XtoY.continuousInterpolator(t1.interpolation, extrapolation);
+
   make(
     `Linear,
-    XYShape.PointwiseCombination.combineLinear(
-      ~fn=(+.),
+    XYShape.PointwiseCombination.combine(
+      (+.),
+      interpolator,
+      interpolator,
       t1.xyShape,
       t2.xyShape,
     ),
     combinedIntegralSum,
+    None,
   );
 };
 
 let toLinear = (t: t): option(t) => {
   switch (t) {
-  | {interpolation: `Stepwise, xyShape, knownIntegralSum} =>
+  | {interpolation: `Stepwise, xyShape, integralSumCache, integralCache} =>
     xyShape
     |> XYShape.Range.stepsToContinuous
-    |> E.O.fmap(make(`Linear, _, knownIntegralSum))
+    |> E.O.fmap(make(`Linear, _, integralSumCache, integralCache))
   | {interpolation: `Linear} => Some(t)
   };
 };
 let shapeFn = (fn, t: t) => t |> getShape |> fn;
-let updateKnownIntegralSum = (knownIntegralSum, t: t): t => {
+
+let updateIntegralSumCache = (integralSumCache, t: t): t => {
+  ...t,
+  integralSumCache,
+};
+
+let updateIntegralCache = (integralCache, t: t): t => {
   ...t,
-  knownIntegralSum,
+  integralCache,
 };
 
 let reduce =
     (
-      ~knownIntegralSumsFn: (float, float) => option(float)=(_, _) => None,
+      ~integralSumCachesFn: (float, float) => option(float)=(_, _) => None,
+      ~integralCachesFn: (t, t) => option(t)=(_, _) => None,
       fn,
       continuousShapes,
     ) =>
   continuousShapes
-  |> E.A.fold_left(combinePointwise(~knownIntegralSumsFn, fn), empty);
+  |> E.A.fold_left(combinePointwise(~integralSumCachesFn, ~integralCachesFn, fn), empty);
 
-let mapY = (~knownIntegralSumFn=_ => None, fn, t: t) => {
-  let u = E.O.bind(_, knownIntegralSumFn);
+let mapY = (~integralSumCacheFn=_ => None,
+            ~integralCacheFn=_ => None,
+            fn, t: t) => {
```

Comment:
Again, I'd name everything, or everything but the last arg even, at this point.
Note that you can do, ```mapY(~fn=_)`` if you need to still